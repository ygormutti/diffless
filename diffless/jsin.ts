import { bind } from 'decko';

interface Constructor {
    new(...args: any[]): {};
}

type Transformer = (key: any, value: any) => any | undefined;

export const DEFAULT_META_KEY = '__JSIN__';

export class BaseJSIN {
    readonly mapping: { [key: string]: Constructor };
    seed: number;

    constructor(
        readonly metaKey: string = DEFAULT_META_KEY,
    ) {
        this.mapping = {};
        this.seed = 0;
    }

    @bind
    enabled(constructor: Constructor) {
        const key = `${constructor.name}_${this.seed++}`;
        (constructor as any)[this.metaKey] = key;
        this.mapping[key] = constructor;
    }

    stringify(obj: any, replacer?: Transformer, space?: string | number): string {
        return JSON.stringify(obj, (key, value) => {
            if (replacer) value = replacer(key, value);
            const constructorKey = value && value.constructor[this.metaKey];
            if (!constructorKey) return value;
            return {
                ...value,
                [this.metaKey]: constructorKey,
            };
        }, space);
    }

    parse(text: string, reviver?: Transformer) {
        return JSON.parse(text, (key, value) => {
            if (value.hasOwnProperty(this.metaKey)) {
                const prototype = this.mapping[value[this.metaKey]].prototype;
                delete value[this.metaKey];
                Object.setPrototypeOf(value, prototype);
            }
            return reviver ? reviver(key, value) : value;
        });
    }
}

export const JSIN = new BaseJSIN();
