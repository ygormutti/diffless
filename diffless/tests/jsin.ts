import { BaseJSIN, DEFAULT_META_KEY } from '../jsin';

const JSIN = new BaseJSIN();

@JSIN.enabled
class Mock {
    constructor(readonly data: {
        readonly str?: string,
        readonly num?: number,
        readonly bool?: boolean,
        readonly arr?: any[],
        readonly obj?: any,
    } = {}) { }

    doesItWork() {
        return true;
    }
}

describe('JSIN', () => {
    it(' enabled should decorate constructor with key', () => {
        const ctor = Mock as any;
        expect(ctor[DEFAULT_META_KEY]).toBe('Mock_0');
    });

    it('stringify should add metadata to instances of serializable constructors', () => {
        const obj = new Mock({
            arr: [new Mock({ obj: null }), true],
            obj: new Mock({ str: 'myString', bool: true }),
        });
        expect(JSIN.stringify(obj, undefined, 4)).toMatchSnapshot();
    });

    it('parse should set prototypes correctly', () => {
        const obj = new Mock();
        const str = JSIN.stringify(obj);

        const parsedObj = JSIN.parse(str);

        expect(parsedObj.doesItWork()).toBe(true);
    });
});
