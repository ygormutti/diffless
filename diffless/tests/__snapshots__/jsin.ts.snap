// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSIN stringify should add metadata to instances of serializable constructors 1`] = `
"{
    \\"data\\": {
        \\"arr\\": [
            {
                \\"data\\": {
                    \\"obj\\": null
                },
                \\"__JSIN__\\": \\"Mock_0\\"
            },
            true
        ],
        \\"obj\\": {
            \\"data\\": {
                \\"str\\": \\"myString\\",
                \\"bool\\": true
            },
            \\"__JSIN__\\": \\"Mock_0\\"
        }
    },
    \\"__JSIN__\\": \\"Mock_0\\"
}"
`;
